ejb/WFLY-17986.adoc

```diff
@@ -0,0 +1,234 @@
+A EAP deployment in OpenShift may be deployed using a non-HA server profile, or using an HA server profile. In the
+former case, the deployment will run on one Pod in the OpenShift cluster. In the latter case, the deployment will
+run on multiple Pods in the OpenShift cluster. In both cases, the external entrypoint to the cluster will be a route
+or an ingress controller endpoint.
+
+=== Features of the EJB client
+The main purpose of the EJB client library is to allow a client application to discover and interact with EJBs
+contained within an EJB-based application deployment on EAP: the client obtains a proxy for the specific bean
+in the deployment it wants to interact with, and uses the proxy to make invocations on the bean.
+However, over and above this basic invocation mechanism, the EJB client provides a number of important features,
+depending on the type of deployment.
+
+The key features for both HA and non-HA deployments include:
+
+* lookup via JNDI or programmatic creation of proxies for EJBs in a deployment
+* client-specific security and discovery configuration
+* transactional scoping of invocations on EJBs in a deployment
+
+Additional features for HA deployments:
+
+* load balancing - requests to SLSBs can be load balanced across the Pods in the EAP deployment's cluster
+* affinity management - requests to SFSBs will be routed to the specific Pod in the EAP cluster on which the session
+was initially created; in other words, the client proxy for the SFSB will have an affinity to a particular backend Pod
+and that affinity will be maintained
+* failover - requests to SFSBs will be retried on a different Pod in the EAP cluster if the Pod processing a request
+fails, and the affinity will be updated
+
+A major part of this issue is to ensure that these key features work in OpenShift as they do in bare metal.
+
+=== The WildFly operator
+The OpenShift environment differs from a bare metal environment in many key ways; for example, in OpenShift, Pod IP
+addresses and storage volume references are transient, whereas on bare metal they are persistent; in OpenShift,
+Pods which fail can be restarted automatically, whereas on bare metal, they are under the control of a system
+administrator; and in OpenShift, the number of instances of Pods backing a service can be scaled up or scaled down
+automatically. In other words, the OpenShift environment is more dynamic than a corresponding bare metal environment.
+
+Unless otherwise constrained, it would be very difficult for an EJB client application to interact with an EAP
+deployment in OpenShift if stricter guarantees were not imposed. For example, discovery of the members of a
+EAP cluster and affinity management of SFSBs would be next to impossible without an assumption persistent Pod
+IP addresses.
+
+The `WildFly operator` can be used to deploy an EAP application in OpenShift and provide the following guarantees:
+
+* persistent IP addresses for Pods
```

Comment:
I think here the correct statement is "persistent name for Pods"

If you try scaling up and down deployments, you get the following as an example:

scale to 3:

wildfly-build-from-server-0		Pod IP: 10.129.3.31	  Host IP: 172.215.0.255
wildfly-build-from-server-1		Pod IP: 10.131.0.9    Host IP: 172.215.1.38
wildfly-build-from-server-2		Pod IP: 10.129.3.32   Host IP: 172.215.0.255

scale to 1 and then back to 3:

wildfly-build-from-server-0		Pod IP: 10.129.3.31	  Host IP: 172.215.0.255 <-- This POD wasn't stopped and IP stays the same
wildfly-build-from-server-1		Pod IP: 10.131.0.10   Host IP: 172.215.1.38  <-- This POD was destroyed and recreated: IP changes
wildfly-build-from-server-2		Pod IP: 10.131.0.11   Host IP: 172.215.1.38  <-- This POD was destroyed and recreated: IP changes

scale to 0 and then to 4:

wildfly-build-from-server-0		Pod IP: 10.129.3.34	  Host IP: 172.215.0.255 <-- This POD was destroyed and recreated: IP changes
wildfly-build-from-server-1		Pod IP: 10.131.0.12   Host IP: 172.215.1.38  <-- This POD was destroyed and recreated: IP changes
wildfly-build-from-server-2		Pod IP: 10.129.3.35   Host IP: 172.215.0.255 <-- This POD was destroyed and recreated: IP changes
wildfly-build-from-server-3		Pod IP: 10.131.0.13   Host IP: 172.215.1.38  <-- This POD was destroyed and recreated: IP changes

Reply:
Yes, good point.  Thanks for providing proof. The wording needs to be changed. I'm wondering how this feature/ guarantee is described in the documentation for stateful sets.